llvm/test/CodeGen/AMDGPU/scale-offset-flat.ll

```diff
@@ -412,11 +411,11 @@ define amdgpu_ps <2 x float> @flat_atomicrmw_b64_rtn_idxprom(ptr align 8 inreg %
 ; GISEL-NEXT:    s_and_not1_saveexec_b32 s0, s2
 ; GISEL-NEXT:    s_cbranch_execz .LBB21_2
 ; GISEL-NEXT:  .LBB21_4: ; %atomicrmw.private
-; GISEL-NEXT:    s_mov_b32 s1, src_flat_scratch_base_lo
-; GISEL-NEXT:    v_cmp_ne_u64_e32 vcc_lo, 0, v[4:5]
 ; GISEL-NEXT:    s_wait_loadcnt_dscnt 0x0
-; GISEL-NEXT:    v_subrev_nc_u32_e32 v0, s1, v4
-; GISEL-NEXT:    s_delay_alu instid0(VALU_DEP_1)
+; GISEL-NEXT:    v_mov_b32_e32 v0, src_flat_scratch_base_lo
+; GISEL-NEXT:    v_cmp_ne_u64_e32 vcc_lo, 0, v[4:5]
+; GISEL-NEXT:    s_delay_alu instid0(VALU_DEP_2) | instskip(NEXT) | instid1(VALU_DEP_1)
+; GISEL-NEXT:    v_sub_nc_u32_e32 v0, v4, v0
```

Comment:
Any idea why v0 didn't get folded into this sub?

Reply:
I think I have some idea. Scalar is not considered acceptable here:
```
llvm-mc -arch=amdgcn -mcpu=gfx1250 <<< 'v_sub_nc_u32_e32 v0, v4, s0'
<stdin>:1:26: error: invalid operand for instruction
v_sub_nc_u32_e32 v0, v4, s0
                         ^
```

Reply:
I.e., only in VOP3 form, but it is already VOP2 at the time of folding.

Reply:
> but it is already VOP2 at the time of folding

Oh, is this in the second run of folding, after shrinking?

Isn't there some logic to commute sub to subrev so that it can be folded?

Anyway this issue does not need to block this patch.

Reply:
And yes, subrev shall be available. Which is a good potential optimization which we do not do there:
```
llvm-mc -arch=amdgcn -mcpu=gfx1250 -show-encoding <<< 'v_subrev_nc_u32 v0, src_flat_scratch_base_lo, v4'
	v_subrev_nc_u32_e32 v0, src_flat_scratch_base_lo, v4 ; encoding: [0xe6,0x08,0x00,0x4e]
```